### PR TITLE
Add support to obtain trust bundle from iotedged or an input file

### PR DESCRIPTION
**Pull request by mrohera**
Add functionality along with tests in the Edge Hub to obtain the trust bundle either from the iotedged or an input file to facilitate development. The trust bundle is not really used here rather this is being staged for additional upcoming features.

### CODE REVIEW

File: edge-util/src/Microsoft.Azure.Devices.Edge.Util/edged/WorkloadClient.cs [hunk 1]
```diff
@@ -1,150 +1,160 @@
...
+
+            using (HttpClient httpClient = HttpClientHelper.GetHttpClient(this.workloadUri))
+            {
+                var edgeletHttpClient = new HttpWorkloadClient(httpClient) { BaseUrl = HttpClientHelper.GetBaseUrl(this.workloadUri) };
+                CertificateResponse result = await this.Execute(() => edgeletHttpClient.CreateServerCertificateAsync(this.apiVersion, this.moduleId, this.moduleGenerationId, request), "CreateServerCertificateAsync");
+                return result;
+            }
+        }
+
+        public async Task<TrustBundleResponse> GetTrustBundleAsync()
```
Comment (mrohera) [inline edge-util/src/Microsoft.Azure.Devices.Edge.Util/edged/WorkloadClient.cs]: Perhaps my editor setting caused the diff here but this function was the only change.

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Service/EdgeHubCertificates.cs [hunk 2]
```diff
@@ -19,10 +19,13 @@ public class EdgeHubCertificates
 
         public IList<X509Certificate2> CertificateChain { get; }
 
-        EdgeHubCertificates(X509Certificate2 serverCertificate, IList<X509Certificate2> certificateChain)
+        public IList<X509Certificate2> TrustBundle { get; }
+
+        EdgeHubCertificates(X509Certificate2 serverCertificate, IList<X509Certificate2> certificateChain, IList<X509Certificate2> trustBundle)
         {
             this.ServerCertificate = Preconditions.CheckNotNull(serverCertificate, nameof(serverCertificate));
             this.CertificateChain = Preconditions.CheckNotNull(certificateChain, nameof(certificateChain));
```
Comment (varunpuranik) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Service/EdgeHubCertificates.cs]: How is the TrustBundle different from the CertificateChain?

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Service/modules/CommonModule.cs [hunk 3]
```diff
@@ -47,7 +49,8 @@ public class CommonModule : Module
             string storagePath,
             Option<string> workloadUri,
             TimeSpan scopeCacheRefreshRate,
-            bool persistTokens)
+            bool persistTokens,
+            IList<X509Certificate2> trustBundle)
```
Comment (varunpuranik) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Service/modules/CommonModule.cs]: Why is this being added here?

File: edge-util/src/Microsoft.Azure.Devices.Edge.Util/CertificateHelper.cs [hunk 4]
```diff
@@ -208,6 +209,12 @@ public static async Task<(X509Certificate2 ServerCertificate, IEnumerable<X509Ce
             return ParseCertificateResponse(response);
         }
 
+        public static async Task<IEnumerable<X509Certificate2>> GetTrustBundleFromEdgelet(Uri workloadUri, string workloadApiVersion, string moduleId, string moduleGenerationId)
```
Comment (varunpuranik) [inline edge-util/src/Microsoft.Azure.Devices.Edge.Util/CertificateHelper.cs]: I am not sure this is a good idea.. The idea for this Helperfile is to contain only static cert related helpers. It should probably not have any workload api/edgelet related APIs.

File: edge-util/src/Microsoft.Azure.Devices.Edge.Util/CertificateHelper.cs [hunk 5]
```diff
@@ -258,6 +265,33 @@ public static IList<string> ParsePemCerts(string pemCerts)
...
+            {
+                certs = sr.ReadToEnd();
+            }
+
+            return ParseTrustedBundleCerts(certs);
+        }
+
+        internal static IEnumerable<X509Certificate2> ParseTrustBundleResponse(TrustBundleResponse response)
+        {
+            return ParseTrustedBundleCerts(response.Certificate);
```
Comment (varunpuranik) [inline edge-util/src/Microsoft.Azure.Devices.Edge.Util/CertificateHelper.cs]: Do we need to check response for null?

File: edge-util/src/Microsoft.Azure.Devices.Edge.Util/CertificateHelper.cs [hunk 6]
```diff
@@ -289,6 +323,10 @@ internal static (X509Certificate2, IEnumerable<X509Certificate2>) ParseCertifica
                 {
                     chain.Add(new X509CertificateEntry(x509Cert));
                 }
+                if (certObject is AsymmetricCipherKeyPair)
+                {
+                    certObject = ((AsymmetricCipherKeyPair)certObject).Private;
```
Comment (varunpuranik) [inline edge-util/src/Microsoft.Azure.Devices.Edge.Util/CertificateHelper.cs]: Curious - what is this change for? Can you please add a comment?

File: edge-util/src/Microsoft.Azure.Devices.Edge.Util/edged/WorkloadClient.cs [hunk 7]
```diff
@@ -1,150 +1,160 @@
...
+        {
+            using (HttpClient httpClient = HttpClientHelper.GetHttpClient(this.workloadUri))
```
Comment (varunpuranik) [inline edge-util/src/Microsoft.Azure.Devices.Edge.Util/edged/WorkloadClient.cs]: Do we need to add a test to WorkloadClientTest?

Comment (mrohera) [reply edge-util/src/Microsoft.Azure.Devices.Edge.Util/edged/WorkloadClient.cs]: Currently is seems like there are no tests.

Comment (varunpuranik) [reply edge-util/src/Microsoft.Azure.Devices.Edge.Util/edged/WorkloadClient.cs]: Emm, I do see some tests in WorkloadClientTest.cs.. Maybe @ancaantochi knows more..

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Service/modules/CommonModule.cs (hunk 3 above)
Comment (mrohera) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Service/modules/CommonModule.cs]: This will need to eventually be passed to the Authenticator which will have the ability to authenticate X.509 certificates. This is specifically needed so that the X.509 CA auth can be done by EdgeHub.

Comment (mrohera) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Service/modules/CommonModule.cs]: Subsequent PRs will use this in short.

File: edge-util/src/Microsoft.Azure.Devices.Edge.Util/CertificateHelper.cs (hunk 5 above)
Comment (varunpuranik) [reply edge-util/src/Microsoft.Azure.Devices.Edge.Util/CertificateHelper.cs]: Nit: Change to expression bodied syntax..

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Service/EdgeHubCertificates.cs (hunk 2 above)
Comment (mrohera) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Service/EdgeHubCertificates.cs]: Trustbundle collection of unrelated root CAs
  **Example:**  FLEET 1 CA | FLEET 2 CA | ...

Certificate Chain is only the Edge Hub server certificate chain. 
  **Example Cert**: Edge Hub Server Cert | INT 2 | INT 1 |  Root CA
  In the example above, the chain would be INT 2 | INT 1 |  Root CA

File: edge-util/src/Microsoft.Azure.Devices.Edge.Util/CertificateHelper.cs (hunk 6 above)
Comment (mrohera) [reply edge-util/src/Microsoft.Azure.Devices.Edge.Util/CertificateHelper.cs]: Sure I can add a comment, I am not exactly sure why this occurred but when testing with PFX certificates generated via openssl app (not dotnet core) the certObject was of type AsymmetricCipherKeyPair and thus its private key needed to accessed differently.

File: edge-util/src/Microsoft.Azure.Devices.Edge.Util/CertificateHelper.cs (hunk 4 above)
Comment (mrohera) [reply edge-util/src/Microsoft.Azure.Devices.Edge.Util/CertificateHelper.cs]: There is prior precedent in this file to obtain the Edge Hub certificate and key. Perhaps this needs to be refactored as a whole. We can do this in a separate PR.

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Service/EdgeHubCertificates.cs (hunk 2 above)
Comment (varunpuranik) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Service/EdgeHubCertificates.cs]: Isn't TrustBundle what is returned to other modules on the Workload API call, and what the modules use to validate the EdgeHub Server Certificate? Or is this something different? In which case, how would the TrustBundle above work?

File: edge-util/src/Microsoft.Azure.Devices.Edge.Util/CertificateHelper.cs (hunk 5 above)
Comment (mrohera) [reply edge-util/src/Microsoft.Azure.Devices.Edge.Util/CertificateHelper.cs]: Yes I can add this. Makes sense.

File: edge-util/src/Microsoft.Azure.Devices.Edge.Util/edged/WorkloadClient.cs (hunk 7 above)
Comment (mrohera) [reply edge-util/src/Microsoft.Azure.Devices.Edge.Util/edged/WorkloadClient.cs]: I am mistaken, I see these tests, will add this specific test as well.

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Service/EdgeHubCertificates.cs (hunk 2 above)
Comment (mrohera) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Service/EdgeHubCertificates.cs]: Yes that is right it is the same TrustBundle that is returned to all modules. All modules including Edge Hub would receive the same set of CA certificates. One among them would be the root of the Edge Hub server certificate.

File: edge-util/src/Microsoft.Azure.Devices.Edge.Util/CertificateHelper.cs (hunk 5 above)
Comment (mrohera) [reply edge-util/src/Microsoft.Azure.Devices.Edge.Util/CertificateHelper.cs]: Added to https://github.com/Azure/iotedge/pull/562

File: edge-util/src/Microsoft.Azure.Devices.Edge.Util/edged/WorkloadClient.cs (hunk 7 above)
Comment (mrohera) [reply edge-util/src/Microsoft.Azure.Devices.Edge.Util/edged/WorkloadClient.cs]: Added in https://github.com/Azure/iotedge/pull/562
